### PR TITLE
Split functions from text.php to new Logger class

### PR DESCRIPTION
**Pull request by zeroadam**
Create Logger class and move logger() and dlogger() from text.php

related to #3878

### CODE REVIEW

File: src/Core/Logger.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+     * LOGGER_TRACE
+     * LOGGER_DEBUG
+     * LOGGER_DATA
+     * LOGGER_ALL
+     *
+     * @global array $LOGGER_LEVELS
+     * @param string $msg
+     * @param int $level
+     */
+    public static function log($msg, $level = LOGGER_INFO)
```
Comment (MrPetovan) [inline src/Core/Logger.php]: `LOGGER_*` constants should be transformed into class constants:
```php
class Logger
{
    const INFO = ...
    const WARNING = ...
```

File: src/Core/Logger.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        $a = get_app();
+        global $LOGGER_LEVELS;
```
Comment (MrPetovan) [inline src/Core/Logger.php]: This `global` variable should be transformed into a static property of this class:
```php
class Logger
{
    ...
    public static $levels = ...
```

File: src/Core/Logger.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                $callers[0]['line'],
+                $callers[1]['function'],
+                $msg
+            );
+
+        $stamp1 = microtime(true);
+        @file_put_contents($logfile, $logline, FILE_APPEND);
+        $a->saveTimestamp($stamp1, "file");
+    }
+}
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Code Standards: Please add a newline at the end of the file.

File: src/Core/Logger.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     * LOGGER_DATA
+     * LOGGER_ALL
+     *
+     * @global array $LOGGER_LEVELS
+     * @param string $msg
+     * @param int $level
+     */
+    public static function log($msg, $level = LOGGER_INFO)
+    {
+        $a = get_app();
```
Comment (nupplaphil) [inline src/Core/Logger.php]: `get_app()` shouldn't be used.

You can use extend Logger from BaseObject like
```php
class Logger extends BaseObject
```
and than use
```php
$a = self::getApp();
```

File: src/Core/Logger.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                if (substr($k, 0, 7) == "LOGGER_") {
+                    $LOGGER_LEVELS[$v] = substr($k, 7, 7);
+                }
+            }
+        }
+
+        $process_id = session_id();
+
+        if ($process_id == '') {
+            $process_id = get_app()->process_id;
```
Comment (nupplaphil) [inline src/Core/Logger.php]: use `$a->process_id`

File: src/Core/Logger.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+        if (count($LOGGER_LEVELS) == 0) {
+            foreach (get_defined_constants() as $k => $v) {
+                if (substr($k, 0, 7) == "LOGGER_") {
+                    $LOGGER_LEVELS[$v] = substr($k, 7, 7);
+                }
+            }
+        }
+
+        $process_id = session_id();
```
Comment (nupplaphil) [inline src/Core/Logger.php]: normally, camelCase is used
`$processId = session_id();`

File: src/Core/Logger.php [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+     * LOGGER_DATA
+     * LOGGER_ALL
+     *
+     * @global array $LOGGER_LEVELS
+     * @param string $msg
+     * @param int $level
+     */
+    public static function devLog($msg, $level = LOGGER_INFO)
+    {
+        $a = get_app();
```
Comment (nupplaphil) [inline src/Core/Logger.php]: see above `$a = self::getApp();`

File: src/Core/Logger.php [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+        if (count($LOGGER_LEVELS) == 0) {
+            foreach (get_defined_constants() as $k => $v) {
+                if (substr($k, 0, 7) == "LOGGER_") {
+                    $LOGGER_LEVELS[$v] = substr($k, 7, 7);
+                }
+            }
+        }
+
+        $process_id = session_id();
```
Comment (nupplaphil) [inline src/Core/Logger.php]: see above `$processId = session_id();`

File: src/Core/Logger.php [hunk 9]
```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * @file src/Core/Logger.php
+ */
+namespace Friendica\Core;
+
+use Friendica\Core\Config;
+use Friendica\Util\DateTimeFormat;
+
+class Logger
```
Comment (rabuzarus) [inline src/Core/Logger.php]: please add a doxygen header for the `class`

File: src/Core/Logger.php [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+{
+    // Log levels:
+    const WARNING = 0;
+    const INFO = 1;
+    const TRACE = 2;
+    const DEBUG = 3;
+    const DATA = 4;
+    const ALL = 5;
+
+    public static $levels = [];
```
Comment (MrPetovan) [inline src/Core/Logger.php]: This array is actually an associative array with the constant value as the key, and the level name as value. You can write it like this:
```php
class Logger 
{
    // Log levels:
    const WARNING = 0;
    const INFO = 1;
    const TRACE = 2;
    const DEBUG = 3;
    const DATA = 4;
    const ALL = 5;
    
     public static $levels = [
         self::WARNING => 'Warning',
         self::INFO => 'Info',
         self::TRACE => 'Trace',
         self::DEBUG => 'Debug',
         self::DATA => 'Data',
         self::ALL => 'All',
     ];
...
```

and get rid of the `getConstants` method.

File: src/Core/Logger.php [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+use Friendica\Core\Config;
+use Friendica\Util\DateTimeFormat;
+use ReflectionClass;
+
+/**
+ * @brief Logger functions
+ */
+class Logger extends BaseObject
+{
+    // Log levels:
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Code standards: Please use tabs for indenting.

File: src/Core/Logger.php [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        if (!is_null($dlogip) && $_SERVER['REMOTE_ADDR'] != $dlogip)
+        {
+            return;
+        }
+
+        if (count(self::$levels) == 0)
+        {
+            foreach (self::getConstants() as $k => $v)
+            {
+                $levels[$v] = $k;
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Neither `self::$levels` not `$levels` are used in this method.

**Comment by MrPetovan** [comment]
Waiting for the tests to complete.

**Comment by nupplaphil** [comment]
@MrPetovan what about the idea with the instance? Because we have to take care that it isn't called before `App`. I think `App` is something like an container for globally needed instances :-)

**Comment by MrPetovan** [comment]
There used to be a race condition where App instances would be infinitely created if there was a reference to the App object before the `App->__construct()` finished. With the addition of `BaseObject::setApp()` at the start of `App->__construct()`
https://github.com/friendica/friendica/blob/f7147fae960db05d75ec86479803e56dbf4f835b/src/App.php#L184-L191
It isn't a concern anymore.

**Comment by nupplaphil** [comment]
hm.. ok..

**Comment by MrPetovan** [comment]
Your idea isn't bad, but it isn't warranted to solve the infinite loop.

**Comment by tobiasd** [comment]
With this merged (2f28c2ebbfd8f66093d7081d40b1d48f65fcff20) I cannot post any new top level postings. Had to revert to f7147fae960db05d75ec86479803e56dbf4f835b (Merge 6041) to get a working not back. Nothing in the Friendica/php logs to find so far.

**Comment by tobiasd** [comment]
Patch for the friendica-projects addon is on its way to @fabrixxm - It would be really great if you could send a bing about such massive changes to the developers forum as not everybody is following all the PRs on github but may be following the that forum.

**Comment by zeroadam** [comment]
@tobiasd I will be more aware of that, and post to the developers forum as I'm doing these PRs, sorry about that.

Also, are you saying that your errors were caused by the projects addon? That you weren't able to make posts?

**Comment by tobiasd** [comment]
Yes, seems so that the moved logger function that was not changed in the projects addon caused an error 500 on making new postings in general.

**Comment by MrPetovan** [comment]
The project addon is a non-official addon maintained by @fabrixxm , hence the compatibility break. You did good on the official repo, and like @tobiasd said, you could do even better by notifying Friendica Developers for breaking changes.

**Comment by MrPetovan** [comment]
After pulling this morning:
- I can't submit top-level posts (without the unofficial project addon).
- I can't follow new people, clicking on the Submit button on the follow page yield a 500 error.

I can't access my error log yet, I'll have a full report tonight, but I'm guessing this is related to this PR.

**Comment by MrPetovan** [comment]
Additional info: My top-level post with tags didn't go through while text-only did, so it might be caused by #6048 instead.

**Pull request merged**